### PR TITLE
Fixed wrong metal addition result

### PR DESCRIPTION
**Pull request by nikviktorovich**
## Summary

This PR fixes a case when `tf2.Metal(1.88) + tf2.Metal(1.11) != tf2.Metal(3)` and adds some more test cases to assure it. It also enforces `Metal.__add__` method to return an instance of a `Metal` instead of a `Fracture` to make addition chains of more than 2 elements stable (due to rounding and returning an instance of `Fracture`, the required precision could be lost after multiple additions). The breaking change here though is adding anything other than `Fracture` with denominator 9 to `Metal` will result in an error.

## Checklist

- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: steam/ext/tf2/currency.py [hunk 1]
```diff
@@ -69,6 +69,15 @@ def __new__(cls, value: SupportsMetal, /, *, _normalize: bool = ...) -> Self:
...
     _denominator: Final = 9
 
+    def __add__(self, other: "Fraction") -> "Metal":
+        if self.denominator != other.denominator:
+            raise ValueError("Unsupported sum due to different denominators")
+
+        numerators_sum = self.numerator + other.numerator
+        precision = 10**2
+        value = int(numerators_sum / self.denominator * precision) / precision
+        return Metal(value)
```
Comment (Gobot1234) [inline steam/ext/tf2/currency.py:79]:
```suggestion
        return Metal(super().__add__(other))
```

**Review by Gobot1234 — CHANGES_REQUESTED** [review]
I agree there are a few issues here I just don't entirely agree with how you've gone about fixing them and there are definitely some floating point rounding issues which I think are better handled in init and the math by the super class's implementation of these methods

**Comment by Gobot1234** [comment]
FWIW I've changed the methods to
```py
    def __new__(cls, value: SupportsMetal, /, *, _normalize: bool = ...) -> Self:
        value = float(value)

        self = object.__new__(cls)

        try:
            true_value = round(value, 2)
        except (ValueError, TypeError):
            raise TypeError("non-int passed to Metal.__new__, that could not be cast") from None

        if not math.isclose(value, true_value):
            raise ValueError("metal value's last digits must be close to 0")

        stred = f"{true_value:.2f}"
        if stred[-1] != stred[-2]:
            raise ValueError("metal value must be a multiple of 0.11")

        fractional, integer = math.modf(true_value)
        self._numerator = int(integer) * 9 + math.floor(fractional * 10)  # I think this is more reliable at extracting the parts here
        return self

	...

    def __str__(self) -> str:
        return f"{math.floor(self._numerator * 100 / 9) / 100:.2f}"  # str(Metal(1.88)) was 1.89 before whoops
```

**Comment by nikviktorovich** [comment]
The problem with your solution is that the result of extracting true value this way
```
fraction = Fraction(<more than 5>, 9)
value = float(fraction)
true_value = round(value, 2)
```
fails both
```
if not math.isclose(value, true_value):
    ...
```
and
```
stred = f"{true_value:.2f}"
if stred[-1] != stred[-2]:
    ...
```
validations in the constructor. But I see the point. I'll try to come up with a different solution, thanks.

**Comment by Gobot1234** [comment]
How does this look? https://gist.github.com/Gobot1234/925ecdaf2d911ccc207f0c77ab62d8b9

**Comment by Gobot1234** [comment]
My issues with actually implementing `__add__` are 1 there are a lot of cases to handle e.g. Fraction() + float() works according to the original implementing, 2 we need to then implement all the other magic methods which is a lot of work, 3 this implementation is actually incompatible with the superclass.

At the end of the day returning a Fraction IMO is fine, it will have all the correct behaviour apart from formatting and if you want that just cast it to Metal.

**Comment by nikviktorovich** [comment]
> How does this look? https://gist.github.com/Gobot1234/925ecdaf2d911ccc207f0c77ab62d8b9 (ignoring the lack of `__add__` overloading) this should allow for infinite precision addition and fix the error where `Metal(1.88) + Metal(1.22) != Metal(3)`

It seems like this way of validation works like a charm!

> 1. there are a lot of cases to handle e.g. `Fraction() + float()` works according to the original implementation but not here

I'm not sure if this should even be allowed. In my opinion, `tf2.Metal(0.11) + tf2.Metal(0.22)` looks much more readable than either `tf2.Metal(0.11) + 0.22` or `tf2.Metal(0.11) + Fraction(2, 9)`. But if you wanna stick with this behavior, there's not really much of a work to implement the magic methods (considering, only addition, subtraction, multiplication and division are allowed):

```python
class Metal(Fraction):
    __slots__ = ()

    @overload
    def __new__(cls, value: SupportsMetal, /) -> Self:  # type: ignore
        ...

    def __new__(cls, value: SupportsMetal, /, *, _normalize: bool = ...) -> Self:
        scrap = cls.extract_scrap(value)
        return super().__new__(cls, scrap, 9)
    
    @classmethod
    def extract_scrap(cls, value: SupportsMetal) -> int:
        if isinstance(value, Fraction):
            if value.denominator in {9, 3, 1}:
                return value.numerator * (9 // value.denominator)
            raise ValueError("cannot convert Fraction to Metal, denominator isn't 1, 3 or 9")

        try:
            value = Decimal(value)
        except (ValueError, TypeError):
            raise TypeError("non-int passed to Metal.__new__, that could not be cast") from None

        fractional, integer = modf(value)
        if not math.isclose(fractional - round(fractional, 2), 0, abs_tol=1e-9):
            raise ValueError("metal value's last digits must be close to 0")

        digits = round(fractional, 2).as_tuple().digits
        if len(digits) >= 2 and digits[0] != digits[1]:
            raise ValueError("metal value must be a multiple of 0.11")
        
        return int(integer) * 9 + digits[0]
    
    def __add__(self, other: SupportsMetal) -> Fraction:
        scrap = Metal.extract_scrap(other)
        return Fraction(self.numerator + scrap, 9)
    
    def __sub__(self, other: SupportsMetal) -> Fraction:
        scrap = Metal.extract_scrap(other)
        return Fraction(self.numerator - scrap, 9)
    
    def __mul__(self, other: SupportsMetal) -> Fraction:
        scrap = Metal.extract_scrap(other)
        return Fraction(self.numerator * scrap, 9)
    
    def __div__(self, other: SupportsMetal) -> Fraction:
        scrap = Metal.extract_scrap(other)

        if self.numerator % scrap != 0:
            raise ValueError(f'{self.numerator} scrap cannot be divided by {scrap}')
        
        return Fraction(scrap, 9)

    def __str__(self) -> str:
        return f"{self.numerator // self.denominator}.{f'{(self.numerator % self.denominator) * 9 // self.denominator}' * 2}"

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({str(self)!r})"
```

What do you think about this implementation?

**Comment by Gobot1234** [comment]
LGMT thanks

**Comment by nikviktorovich** [comment]
Okay then, I changed the PR commit. Thanks.

**Comment by Gobot1234** [comment]
@nikviktorovich how does this all look now?

**Comment by nikviktorovich** [comment]
Tests look good and pass successfully, so it looks like the issue is resolved. Great work!

**Pull request merged**